### PR TITLE
[ONNX] Fix concat with empty tensors

### LINKED ISSUE #54410: I converted pytorch model to onnx, but failed to run onnx.
## 🐛 Bug

I converted pytorch model to onnx, but failed to run onnx. 

## To Reproduce

1. I tried to convert retinanet to onnx model in this git repository: https://github.com/yhenon/pytorch-retinanet/
2. conversion succeed but inference onnx model failed.
3. error logs when inference:
onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running Concat node. Name:'Concat_711' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/tensor/concat.cc:72 onnxruntime::common::Status onnxruntime::ConcatBase::PrepareForCompute(onnxruntime::OpKernelContext*, const std::vector<const onnxruntime::Tensor*>&, onnxruntime::Prepare&) const inputs_n_rank == inputs_0_rank was false. Ranks of input data are different, cannot concatenate them. expected rank: 1 got: 2

4. Here is my conversion/inference code:
```
import torch
import torch.onnx
from retinanet import model

model= model.resnet50(num_classes=80,)
model.load_state_dict(torch.load('../models/coco_resnet_50_map_0_335_state_dict.pt'))
model=model.cuda()
dummy_input = torch.randn(1, 3, 608, 1024, device='cuda')
torch.onnx.export(model, dummy_input, "onnx_model_name.onnx", opset_version=11)

import onnx
model = onnx.load("onnx_model_name.onnx")
onnx.checker.check_model(model)
print(onnx.helper.printable_graph(model.graph))

import onnxruntime as ort
import numpy as np
ort_session = ort.InferenceSession('onnx_model_name.onnx')
input_name = ort_session.get_inputs()[0].name # 'data'
out = ort_session.get_outputs()
outputs = list()
for i in out:
   outputs.append(i.name)
print(input_name)
print(outputs)

outputs = ort_session.run(outputs, input_feed={input_name: np.random.randn(1, 3, 608, 1024).astype(np.float32)})
print(outputs[0])`
```

## Expected behavior

error disappear.

## Environment

Collecting environment information...
PyTorch version: 1.8.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux bullseye/sid (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.18.4

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070 SUPER
Nvidia driver version: 460.56
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.0+cu111
[pip3] torchaudio==0.8.0
[pip3] torchinfo==0.0.8
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.9.0+cu111
[conda] Could not collect

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof @SplitInfinity

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Fixes #54410

### CODE REVIEW

File: torch/onnx/symbolic_opset11.py [hunk 1]
```diff
@@ -537,6 +537,26 @@ def Delete(g: jit_utils.GraphContext, tensor_list, dim):
 @_beartype.beartype
 def cat(g: jit_utils.GraphContext, tensor_list, dim):
     if symbolic_helper._is_packed_list(tensor_list):
+        tensors = symbolic_helper._unpack_list(tensor_list)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py:540]: Duplicated block? I assume `opset9.cat` call below would cover this.

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -529,6 +529,25 @@ def reciprocal(g: jit_utils.GraphContext, self):
 @_beartype.beartype
 def cat(g: jit_utils.GraphContext, tensor_list, dim):
     tensors = symbolic_helper._unpack_list(tensor_list)
+    # torch.cat ignores empty tensors such as `torch.Tensor([])`
+    # These needs to be removed as input from ONNX's concat too, otherwise shape inference
+    # will likely fail due to inputs with different ranks (0 for empty tensor, > 0 for anything else)
+    nonempty_tensors = []
+    for t in tensors:
+        if symbolic_helper._is_constant(t) and not symbolic_helper._get_tensor_dim_size(t, 0):
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:537]: Should tensors with 0 in their dimension size count as "empty tensor" under this context? For example `torch.ones(2, 0, 3)`. (Maybe create a helper function `is_empty_tensor`)

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -529,6 +529,25 @@ def reciprocal(g: jit_utils.GraphContext, self):
...
+            continue
+        nonempty_tensors.append(t)
+    assert len(nonempty_tensors) > 0
+    assert all(
+        [
+            symbolic_helper._get_tensor_rank(t)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:543]: `_get_tensor_rank` returns None when rank is unknown. Is it expected to raise assertion error for this case?

Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset9.py]: This assert only makes sense if `symbolic_helper._get_tensor_rank(nonempty_tensors[0] is not None`, so I will change it accordingly

File: torch/onnx/symbolic_opset9.py (hunk 2 above)
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset9.py]: That is a good question. How would we detect one scenario or the other? both would have value `0`

**Review by BowenBao — APPROVED** [review]

**Pull request closed without merging**